### ISSUE TITLE
Ergonomic issues caused by swallowing error conditions

### ISSUE THREAD

**Issue by MrJoy** [avo-hq/avo#785]
### Describe the bug

There's a few cases where Avo either swallows errors or fails to check things strictly and the end result is misleading behavior.

### Steps to Reproduce

1. If you have a typo in the named parameters for `field`, you can get surprising behavior.  Checking for unexpected/unknown parameters would be helpful here.  E.G. I accidentally typed `hide_in` and was baffled by why the field kept showing up.
2. If you provide a non-existent property/method name to `field`, you simply don't get a field.
3. If you provide a field type that doesn't exist, you are simply not shown the field.

```ruby
field :some_field_that_exists, as: :text, hide_in: [:index]
field :some_field_that_exists, as: :some_type_that_doesnt_exist
field :some_field_that_doesnt_exist, as: :text
```

### Expected behavior

Timely, in-my-face errors informing me that Avo has NFI what I mean would be super helpful.

### Actual behavior

Avo seems to try to avoid embarrassing me by pretending I didn't say what I said.

### System configuration

**Avo version**: 2.1.1

**Rails version**: 7.0.2.3

**Ruby version**: 3.0.3

#### Impact

 - [ ] High impact
 - [ ] Medium impact
 - [x] Low impact

#### Urgency

 - [ ] High urgency
 - [ ] Medium urgency
 - [x] low urgency

**Comment by adrianthedev** [comment]
Yeah, I know what you mean. 

The way it works is that it takes the args dynamically, and if they are recognized by the field, they will get parsed accordingly. If their options are wrong, that might trigger an error. When an option is unrecognized it's just ignored.
When the `as:` option is invalid it will trigger a (very obscure) terminal message (this can be improved).
When you pass in a db property that doesn't exist, the field will be treated with the value as being `nil` so it will display a dash on screen `–`.


The output for the invalid field scenario (from your example)👇

```
web    | [Avo] The invalid field, as: invalid_field from TeamResource has an invalid configuration.
```

<img width="503" alt="CleanShot 2022-04-05 at 11 06 39@2x" src="https://user-images.githubusercontent.com/1334409/161719805-3661c39b-69cb-435e-a6f6-6f38a5205ff6.png">


To be able to give errors regarding typos like `hide_in` -> `hide_on`, we'd have to rearchitect the fields a bit. At the moment there isn't a place where we specify the options that a field can and can't take on. As I explained it's a "give me all args and I'll pick the ones I need" game. 
Most options are on `BaseField` and some of them don't apply to all of them (by design). Like `required` and `sortable` options don't get applied to `has_many` fields.

I agree that this is something that should be improved, but I'd like to do it when some of the more important features are taken care of like sidebar editor, resource tools, settings screens and others.
If you have any ideas on how the invalid arguments (the ones below) can be easily fixed, please let me know, I'd love to spitball ideas.

```ruby
field :some_field_that_exists, as: :text, hide_in: [:index]
field :some_field_that_exists, as: :some_type_that_doesnt_exist
```

For the third one, I created [a ticket](https://github.com/avo-hq/avo/issues/788) to surface that issue better.

```ruby
field :some_field_that_doesnt_exist, as: :text
```

**Comment by MrJoy** [comment]
I think fixing it will probably require adjustments to how field types are defined, for sure.  And, unfortunately, that will almost certainly wind up being a breaking change for custom field types.  As such, leaning into it sooner while the ecosystem is small improves the odds that you can fix it without disrupting project momentum.  Alternatively, leaving this as a chronic pain-point lets you focus on pre-established priorities but has an (admittedly unmeasurable) risk of adding some friction against adoption.  It's a tough spot to be in.

If the field class maintained some introspectable state about what options it accepted, then you could do a parameters check when `field` is called from a resource class.  Right now you're using `attr_reader` as a convention for fields to expose configuration to downstream resources.  Perhaps an approach in which one defines parameters using a DSL construct here, and Avo takes care of populating those attributes instead of passing in `**args` to the constructor would be a viable approach?  You'd still need to have some post-initialize-pass so the Field could apply whatever adjustments are needed beyond simple default values, but I think that would give a Pareto-optimal win of covering the most common cases, and leaving more esoteric situations to be handled by the particular Field implementation as-needed.

Does that make sense?

**Comment by MrJoy** [comment]
... as for the non-existent field showing as blank:  I must've also had an invalid field type or something, because I wasn't seeing the column/row _at all_.  D'oh.  My mistake.

**Comment by adrianthedev** [comment]
Yes. All you said above makes total sense. Thanks for the elaborate message. The DSL approach is the right path to take and we'll address that when we start to work on our plugins system.

I agree it's a pain point right now, but not a huge issue. It just makes the developer experience a bit more cumbersome when passing invalid options.

There are tons of state-of-the-art tech projects in the cemetery. They had the perfect CI/CD pipeline, the best typed system, followed all the best practices, etc., but haven't been able to prove themselves useful.
At this point, I'd rather make sure that developers can create the best experiences for their teams using Avo. There are still a lot of things that are missing from making Avo a wholesome product to help devs do that.
Once you have all the tools to create something complete, we can focus more on these types of things. You might say we're still validating the product and that means that some features like these will be pushed to the periphery a bit.

Does that make sense as our process going forward?

**Comment by MrJoy** [comment]
I agree that prioritizing developer productivity and getting to results is crucial.  I tend to view this in terms of that, rather than as a larger "correctness" or "rigor" situation.

One of the things I've found that's left me dissatisfied with many developer tools / libraries are little pain-points that result in a disproportionate amount of lost time.  I suspect I'm more sensitive to this because of my ADHD, but I like to think it makes me a good canary-in-the-coal-mine for ergonomic issues with developer tools.

A good example of this for me, is Arbre in ActiveAdmin.  Because my work in ActiveAdmin tends to extremely burst-y (setting up a new project / creating a new product), with small patches of incremental work that tend to not involve writing much new Arbre code, I wind up forgetting that Arbre uses implicit returns from blocks for providing child content to a tag/component in at least some circumstances.  I invariably wind up losing some time wondering why I'm only seeing part of my component, and trying to figure out what I did wrong.  Of course, if I were working with it every day that would get baked into my brain and it would just be a tiny learning curve thing.

Or, when HomeBrew removes a Cask that I have installed, then changes the DSL in a breaking way, and then everything breaks on my machine because the old Cask definition is lingering and being parsed on each run.  There's a particular way to purge the data but I use it so infrequently (a couple times a year, at most) that I never have it at hand and always spend an hour tracking it down (or tracking down the latest incantation to fix the mess).

So, y'know consider me an especially squeeky wheel, but my motivation when making a suggestion will always be about things that I think will help with my productivity when using the tool, the reliability/robustness of it in practical/real terms, etc.  I've been a purist in the past, and that's just too exhausting and usually counter-productive.

That said:  I completely understand that it isn't a priority.  I just wanted to bring it up early so you could make an informed decision about whether it's better to lean into it to head off the risk that fixing it becomes a breaking change that slows adoption of future versions vs. postponing it to prioritize finding product-market fit before money becomes an issue.  You know your situation better than I do, so I'm not going to second guess your judgement.
